src/collapse/xctrace.rs

```diff
@@ -0,0 +1,455 @@
+use quick_xml::{
+    events::{attributes::Attributes, Event},
+    reader::Reader,
+};
+use std::{
+    borrow::Cow,
+    collections::BTreeMap,
+    io::{self, BufRead},
+    sync::Arc,
+};
+
+use super::{
+    common::{fix_partially_demangled_rust_symbol, Occurrences},
+    Collapse,
+};
+
+const REF: &[u8] = b"ref";
+const ID: &[u8] = b"id";
+const NAME: &[u8] = b"name";
+const TRACE_QUERY_RESULT: &[u8] = b"trace-query-result";
+const NODE: &[u8] = b"node";
+const ROW: &[u8] = b"row";
+const BACKTRACE: &[u8] = b"backtrace";
+const FRAME: &[u8] = b"frame";
+
+/// Context of collapsing a xctrace's `Time Profiler` xml
+#[derive(Default)]
+struct CollapseContext {
+    /// xml tag backtrace
+    state_backtrace: TagBacktrace,
+    // --------- per-xml caches below -----------
+    /// backtrace_id <--> BackTrace
+    backtraces: BTreeMap<u64, Arc<Backtrace>>,
+    /// backtrace_id <--> Frame
+    frames: BTreeMap<u64, Arc<Frame>>,
+}
+
+// Note that sometimes same Backtrace have different id, we need to merge them
+// before writing folded file.
```

Comment:
Do you know why this happens? It could be the kind of thing that's needed to support over-time mode ("flamechart mode") perhaps, I'm not sure.